onnxruntime/core/providers/dml/DmlExecutionProvider/src/TensorDesc.cpp

```diff
@@ -315,3 +315,52 @@
     }
     m_bufferTensorDesc.DimensionCount = newDimensionCount;
 }
+
+// Uses dimensionMapping to reorder m_sizes and m_strides to match specific Tensor layout
+void TensorDesc::PermuteDimensions(const std::vector<uint32_t> dimensionMapping, const TensorAxis alignment)
+{
+    EnsureMinimumDimensionCount(dimensionMapping.size(), alignment);
+    InitializeStrides(static_cast<int>(dimensionMapping.size()), alignment);
+    PermuteArray(dimensionMapping, alignment);
+}
+
+// Shuffle m_sizes and m_strides acording to the indexes pointed by dimensionMapping
+void TensorDesc::PermuteArray(const std::vector<uint32_t> dimensionMapping, const TensorAxis alignment)
+{
+    std::vector<uint32_t> temp_sizes(dimensionMapping.size());
+    std::vector<uint32_t> temp_strides(dimensionMapping.size());
+    // Right alignment values are shifted to the end
+    int alignmentOffset = (alignment == TensorAxis::RightAligned) ? MaximumDimensionCount - static_cast<int>(dimensionMapping.size()) : 0;
+
+    for (size_t i = 0; i < dimensionMapping.size(); i++) {
+        temp_sizes[i] = m_sizes[dimensionMapping[i] + alignmentOffset];
+        temp_strides[i] = m_strides[dimensionMapping[i] + alignmentOffset];
+    }
+
+    std::copy(temp_sizes.begin(), temp_sizes.end(), m_sizes + alignmentOffset);
+    std::copy(temp_strides.begin(), temp_strides.end(), m_strides + alignmentOffset);
+    m_bufferTensorDesc.Sizes = m_sizes;
```

Comment:
The 2nd copy can be elided by copying the values in the constructor, rather than a resize with empty values followed by a copy. e.g.

`std::vector<uint32_t> oldValues(values);`